File: sdks/ts/memory/src/acl/wrap.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * `withAccessControl(store, acl, subject, opts)` - wrap a `Store` with an
 * authorisation check on every read/write surface.
 *
 * Each `Store` method is mapped to an `Action`:
 *
 *   read, exists, stat, list, batch-read, localPath  -> 'read'
 *   write, append, batch-write, batch-append         -> 'write'
 *   delete, batch-delete                             -> 'delete'
 *   rename, batch-rename                             -> 'write' (both sides)
 *
 * The resource targeted by a given path is resolved via
 * `opts.resolveResource`. If not supplied, we fall back to `opts.resource`,
 * which covers the common case of one `Store` per brain/collection.
 */

import type {
  Batch,
  BatchOptions,
  EventSink,
  FileInfo,
  ListOpts,
  Path,
  Store,
  Unsubscribe,
} from '../store/index.js'
import type { AccessControlProvider, Action, Resource, Subject } from './index.js'
import { ForbiddenError } from './index.js'

export type WithAclOptions = {
  readonly resource?: Resource
  readonly resolveResource?: (path: Path) => Resource
}

export const withAccessControl = (
  store: Store,
  acl: AccessControlProvider,
  subject: Subject,
  opts: WithAclOptions = {},
): Store => {
  const resolve = (path: Path): Resource => {
    if (opts.resolveResource) return opts.resolveResource(path)
    if (opts.resource) return opts.resource
    throw new ForbiddenError(
      subject,
      'read',
      { type: 'document', id: path },
      'no resource resolver configured',
    )
  }

  const guard = async (action: Action, path: Path): Promise<void> => {
    const resource = resolve(path)
    const result = await acl.check(subject, action, resource)
    if (!result.allowed) {
      throw new ForbiddenError(subject, action, resource, result.reason)
    }
  }

  const wrappedBatch = (batch: Batch): Batch => ({
    read: async (p) => {
      await guard('read', p)
      return batch.read(p)
    },
    write: async (p, c) => {
      await guard('write', p)
      return batch.write(p, c)
    },
    append: async (p, c) => {
      await guard('write', p)
      return batch.append(p, c)
    },
    delete: async (p) => {
      await guard('delete', p)
      return batch.delete(p)
    },
    rename: async (src, dst) => {
      await guard('write', src)
      await guard('write', dst)
      return batch.rename(src, dst)
    },
    exists: async (p) => {
      await guard('read', p)
      return batch.exists(p)
    },
    stat: async (p) => {
      await guard('read', p)
      return batch.stat(p)
    },
    list: async (dir, listOpts) => {
      if (dir !== '') await guard('read', dir)
      return batch.list(dir, listOpts)
    },
  })

  return {
    read: async (path: Path): Promise<Buffer> => {
      await guard('read', path)
      return store.read(path)
    },
    write: async (path: Path, content: Buffer): Promise<void> => {
      await guard('write', path)
      return store.write(path, content)
    },
    append: async (path: Path, content: Buffer): Promise<void> => {
      await guard('write', path)
      return store.append(path, content)
    },
    delete: async (path: Path): Promise<void> => {
      await guard('delete', path)
      return store.delete(path)
    },
    rename: async (src: Path, dst: Path): Promise<void> => {
      await guard('write', src)
      await guard('write', dst)
      return store.rename(src, dst)
    },
    exists: async (path: Path): Promise<boolean> => {
      await guard('read', path)
      return store.exists(path)
    },
    stat: async (path: Path): Promise<FileInfo> => {
      await guard('read', path)
      return store.stat(path)
    },
    list: async (dir: Path | '', listOpts?: ListOpts): Promise<FileInfo[]> => {
      if (dir !== '') await guard('read', dir)
      return store.list(dir, listOpts)
    },
    batch: async (bopts: BatchOptions, fn: (batch: Batch) => Promise<void>): Promise<void> => {
      return store.batch(bopts, (inner) => fn(wrappedBatch(inner)))
    },
    subscribe: (sink: EventSink): Unsubscribe => store.subscribe(sink),
    // `localPath` cannot enforce ACL safely because the provider contract is
    // async. Return no path at all rather than leak a filesystem location.
    localPath: (_path: Path): string | undefined => undefined,
    close: (): Promise<void> => store.close(),
  }
}

File: sdks/py/src/jeffs_brain_memory/acl/wrap.py
# SPDX-License-Identifier: Apache-2.0
"""`wrap_store(store, provider, subject, ...)` - guard a `Store` with ACL.

Each `Store` method maps to an `Action`:

  read, exists, stat, list, batch-read         -> 'read'
  write, append, batch-write, batch-append     -> 'write'
  delete, batch-delete                         -> 'delete'
  rename, batch-rename                         -> 'write' (both sides)

The resource targeted by a path is resolved via `resolve_resource` if
provided; otherwise the wrapper falls back to a fixed `resource`. When
neither is configured the wrapper raises `ForbiddenError` rather than
let the call through unchecked.
"""

from __future__ import annotations

from typing import AsyncIterator, Callable

from ..path import BrainPath
from ..store import (
    Batch,
    BatchOptions,
    ChangeEvent,
    FileInfo,
    ListOpts,
    Store,
)
from .types import Action, ForbiddenError, Provider, Resource, Subject

__all__ = ["wrap_store"]


def wrap_store(
    store: Store,
    provider: Provider,
    subject: Subject,
    *,
    resource: Resource | None = None,
    resolve_resource: Callable[[BrainPath], Resource] | None = None,
) -> Store:
    """Return a `Store` that runs an ACL check before every delegated call."""
    return _WrappedStore(
        store=store,
        provider=provider,
        subject=subject,
        resource=resource,
        resolve_resource=resolve_resource,
    )


class _WrappedStore(Store):
    def __init__(
        self,
        *,
        store: Store,
        provider: Provider,
        subject: Subject,
        resource: Resource | None,
        resolve_resource: Callable[[BrainPath], Resource] | None,
    ) -> None:
        self._store = store
        self._provider = provider
        self._subject = subject
        self._resource = resource
        self._resolve_resource = resolve_resource

    def _resolve(self, path: BrainPath | str) -> Resource:
        bp = BrainPath(str(path))
        if self._resolve_resource is not None:
            return self._resolve_resource(bp)
        if self._resource is not None:
            return self._resource
        raise ForbiddenError(
            self._subject,
            "read",
            Resource(type="document", id=str(path)),
            "no resource resolver configured",
        )

    async def _guard(self, action: Action, path: BrainPath | str) -> None:
        resource = self._resolve(path)
        result = await self._provider.check(self._subject, action, resource)
        if not result.allowed:
            raise ForbiddenError(self._subject, action, resource, result.reason)

    # --- read side -------------------------------------------------------

    async def read(self, path: BrainPath) -> bytes:
        await self._guard("read", path)
        return await self._store.read(path)

    async def exists(self, path: BrainPath) -> bool:
        await self._guard("read", path)
        return await self._store.exists(path)

    async def stat(self, path: BrainPath) -> FileInfo:
        await self._guard("read", path)
        return await self._store.stat(path)

    async def list(
        self,
        dir: BrainPath | str = "",
        opts: ListOpts | None = None,
    ) -> list[FileInfo]:
        if str(dir) != "":
            await self._guard("read", dir)
        return await self._store.list(dir, opts)

    # --- write side ------------------------------------------------------

    async def write(self, path: BrainPath, content: bytes) -> None:
        await self._guard("write", path)
        await self._store.write(path, content)

    async def append(self, path: BrainPath, content: bytes) -> None:
        await self._guard("write", path)
        await self._store.append(path, content)

    async def delete(self, path: BrainPath) -> None:
        await self._guard("delete", path)
        await self._store.delete(path)

    async def rename(self, src: BrainPath, dst: BrainPath) -> None:
        await self._guard("write", src)
        await self._guard("write", dst)
        await self._store.rename(src, dst)

    # --- transactional / subscriber surface ------------------------------

    async def batch(
        self,
        fn: Callable[[Batch], object],
        opts: BatchOptions | None = None,
    ) -> None:
        async def runner(inner: Batch) -> None:
            wrapped = _WrappedBatch(inner, self)
            result = fn(wrapped)
            if hasattr(result, "__await__"):
                await result

        await self._store.batch(runner, opts)

    def subscribe(self, sink: Callable[[ChangeEvent], None]) -> Callable[[], None]:
        return self._store.subscribe(sink)

    def events(self) -> AsyncIterator[ChangeEvent]:
        return self._store.events()

    async def close(self) -> None:
        await self._store.close()

    def local_path(self, path: BrainPath) -> str | None:
        # `local_path` cannot enforce ACL safely because the provider contract
        # is async. Return no path at all rather than leak a filesystem location.
        return None


class _WrappedBatch(Batch):
    """Per-op guarded view onto an underlying `Batch`."""

    def __init__(self, inner: Batch, parent: _WrappedStore) -> None:
        self._inner = inner
        self._parent = parent

    async def read(self, path: BrainPath) -> bytes:
        await self._parent._guard("read", path)
        return await self._inner.read(path)

    async def exists(self, path: BrainPath) -> bool:
        await self._parent._guard("read", path)
        return await self._inner.exists(path)

    async def stat(self, path: BrainPath) -> FileInfo:
        await self._parent._guard("read", path)
        return await self._inner.stat(path)

    async def list(
        self,
        dir: BrainPath | str = "",
        opts: ListOpts | None = None,
    ) -> list[FileInfo]:
        if str(dir) != "":
            await self._parent._guard("read", dir)
        return await self._inner.list(dir, opts)

    async def write(self, path: BrainPath, content: bytes) -> None:
        await self._parent._guard("write", path)
        await self._inner.write(path, content)

    async def append(self, path: BrainPath, content: bytes) -> None:
        await self._parent._guard("write", path)
        await self._inner.append(path, content)

    async def delete(self, path: BrainPath) -> None:
        await self._parent._guard("delete", path)
        await self._inner.delete(path)

    async def rename(self, src: BrainPath, dst: BrainPath) -> None:
        await self._parent._guard("write", src)
        await self._parent._guard("write", dst)
        await self._inner.rename(src, dst)
